File: qfield/charge.py
import pygame
import numpy as np
import draw
from colors import RED, BLUE


class Charge:
    ELEMENTAL_CHARGE = 1.602e-19  # C
    K = 8.988e9  # Nm^2/C^2
    PROTON_MASS = 1.673e-27  # kg
    SCALE = 1e-9  # pixel to nm
    RADIUS = 10  # render purposes only
    HEIGHT = 9e-9  # height difference between fixed and free particles

    # SABA(10,6,4) symplectic integrator coefficients
    c = np.array(
        [
            0.03809449742241219545697532230863756534060,  # c0
            0.1452987161169137492940200726606637497442,  # c1
            0.2076276957255412507162056113249882065158,  # c2
            0.4359097036515261592231548624010651844006,  # c3
            -0.6538612258327867093807117373907094120024,  # c4
            0.4359097036515261592231548624010651844006,  # c3
            0.2076276957255412507162056113249882065158,  # c2
            0.1452987161169137492940200726606637497442,  # c1
        ]
    )
    d = np.array(
        [
            0.09585888083707521061077150377145884776921,  # d0
            0.2044461531429987806805077839164344779763,  # d1
            0.2170703479789911017143385924306336714532,  # d2
            -0.01737538195906509300561788011852699719871,  # d3
            -0.01737538195906509300561788011852699719871,  # d3
            0.2170703479789911017143385924306336714532,  # d2
            0.2044461531429987806805077839164344779763,  # d1
            0.09585888083707521061077150377145884776921,  # d0
        ]
    )
    pos = 0

    def __init__(self, x, y, charge, fixed, mass=1.0):
        self.color = RED if charge > 0 else BLUE
        self.charge = charge * self.ELEMENTAL_CHARGE
        self.mass = mass * self.PROTON_MASS
        self.fixed = fixed
        self.position = np.array([float(x), float(y)])
        self.velocity = np.array([0.0, 0.0])
        self.acceleration = np.array([0.0, 0.0])
        self.force = np.array([0.0, 0.0])

    def apply_force(self, force):
        """Newton's Second Law"""
        self.force = force
        self.acceleration = force / self.mass

    def update(self, time_step):
        """Update position using SABA(10,6,4) symplectic integration"""
        if self.fixed:
            return
        self.velocity += self.acceleration * time_step * self.d[self.pos]
        self.position += self.velocity * time_step * self.c[self.pos]
        self.pos += 1
        self.pos %= 8

    def superposition(self, charges):
        """Calculate superposition of electrostatic forces"""
        total_force = np.array([0.0, 0.0])
        for charge in charges:
            total_force += self.electrostatic_force(charge)
        return total_force

    def electrostatic_force(self, other):
        """Calculate electrostatic force with Coulomb's Law"""
        r = (self.position - other.position) * self.SCALE
        r_magnitude = np.linalg.norm(r)
        force_magnitude = (
            self.K
            * (self.charge * other.charge * r_magnitude)
            / (r_magnitude**2 + self.HEIGHT**2) ** (3 / 2)
        )
        force = force_magnitude * (r / r_magnitude)
        return force

    def render(self, screen):
        """Render charge"""
        pygame.draw.circle(
            screen,
            self.color,
            (int(self.position[0]), int(self.position[1])),
            self.RADIUS,
        )

    def render_velocity(self, screen):
        end = self.position + self.velocity / self.SCALE / 1e15
        draw.draw_arrow(screen, self.position, end, "blue")

    def render_force(self, screen):
        end = (
            self.position
            + self.force / self.SCALE * screen.get_size() * screen.get_size()
        )
        draw.draw_arrow(screen, self.position, end, "red")

File: java/core/src/main/java/io/github/hey2022/qfield/Charge.java
package io.github.hey2022.qfield;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.math.Circle;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.Array;
import space.earlygrey.shapedrawer.ShapeDrawer;

public class Charge {
  private static final float[] c = {
    0.03809449742241219545697532230863756534060f,
    0.1452987161169137492940200726606637497442f,
    0.2076276957255412507162056113249882065158f,
    0.4359097036515261592231548624010651844006f,
    -0.6538612258327867093807117373907094120024f,
    0.4359097036515261592231548624010651844006f,
    0.2076276957255412507162056113249882065158f,
    0.1452987161169137492940200726606637497442f,
  };
  private static final float[] d = {
    0.09585888083707521061077150377145884776921f,
    0.2044461531429987806805077839164344779763f,
    0.2170703479789911017143385924306336714532f,
    -0.01737538195906509300561788011852699719871f,
    -0.01737538195906509300561788011852699719871f,
    0.2170703479789911017143385924306336714532f,
    0.2044461531429987806805077839164344779763f,
    0.09585888083707521061077150377145884776921f,
  };
  private final float ELEMENTAL_CHARGE = 1.602e-19f;
  private final float K = 8.988e9f;
  private final float PROTON_MASS = 1.673e-27f;
  public static final float RADIUS = 10;
  private static final float SELECTED_LIGHT_FACTOR = 1 / 1.5f;
  private final float HEIGHT = 9e-6f;
  private final float HEIGHT2 = (float) Math.pow(HEIGHT, 2);
  private final int PATH_LENGTH = 10000;

  private Vector2 position;
  private Vector2 screenPosition;
  private Vector2 force;
  private Vector2 acceleration;
  private Vector2 velocity;
  private float mass;
  private float distanceTraveled;
  private boolean fixed;
  private Color color;
  private float charge;
  private Path path;
  private boolean selected;
  public Circle circle;
  public boolean drawArrow = true;
  public boolean drawPath = true;

  public Charge(float x, float y, float charge, boolean fixed, float mass) {
    this.mass = mass * this.PROTON_MASS;
    this.fixed = fixed;
    this.color = charge > 0 ? Color.RED : Color.BLUE;
    this.charge = charge * this.ELEMENTAL_CHARGE;
    this.reset(x, y);
    this.selected = false;
  }

  public void reset(float x, float y) {
    position = new Vector2(x, y);
    screenPosition = position.cpy().scl(1 / Main.SCALE);
    circle = new Circle(screenPosition, RADIUS);
    velocity = new Vector2(0, 0);
    acceleration = new Vector2(0, 0);
    force = new Vector2(0, 0);
    path = new Path(PATH_LENGTH);
    distanceTraveled = 0.0f;
  }

  public void applyForce(Vector2 force) {
    this.force = force.cpy();
    this.acceleration = force.cpy().scl(1f / this.mass);
  }

  public void updateForce(Array<Charge> charges) {
    applyForce(superposition(charges));
  }

  public void update(Array<Charge> charges, float timeStep) {
    if (fixed) return;
    Vector2 initialPosition = position.cpy();
    for (int i = 0; i < 8; i++) {
      updateForce(charges);
      velocity.add(acceleration.cpy().scl(timeStep * d[i]));
      position.add(velocity.cpy().scl(timeStep * c[i]));
    }
    distanceTraveled += initialPosition.sub(position).len();
    screenPosition = position.cpy().scl(1 / Main.SCALE);
    circle.setPosition(screenPosition);
    path.add(screenPosition);
  }

  public float getCharge() {
    return charge;
  }

  public Vector2 getPos() {
    return position;
  }

  public Vector2 getScreenPos() {
    return screenPosition;
  }

  public void setPos(Vector2 pos) {
    this.position = pos;
  }

  public boolean isSelected() {
    return selected;
  }

  public void select() {
    this.selected = true;
  }

  public void unselect() {
    this.selected = false;
  }

  public Vector2 superposition(Array<Charge> charges) {
    if (charges == null || charges.size == 0) {
      return new Vector2(0, 0);
    }

    Vector2 force = new Vector2(0, 0);
    for (int i = 0; i < charges.size; i++) {
      Charge charge = charges.get(i);
      Vector2 r = position.cpy().sub(charge.getPos());
      float r2 = r.len2();
      if (r2 == 0) continue; // Skip self interaction
      float f =
          (K * charge.getCharge() * this.charge * (float) Math.sqrt(r2))
              / (float) Math.pow((r2 + HEIGHT2), 1.5);
      force.add(r.nor().scl(f));
    }
    return force;
  }

  public float energy(Array<Charge> charges) {
    return kineticEnergy() + electricPotential(charges);
  }

  public float kineticEnergy() {
    float kineticEnergy = (float) (1.0 / 2.0 * mass * velocity.len2());
    return kineticEnergy;
  }

  public float electricPotential(Array<Charge> charges) {
    float electricPotential = 0.0f;
    if (charges == null) {
      return electricPotential;
    }

    float HEIGHT2 = (float) Math.pow(HEIGHT, 2);
    for (int i = 0; i < charges.size; i++) {
      Charge charge = charges.get(i);
      Vector2 r = position.cpy().sub(charge.getPos());
      float r2 = r.len2();
      if (r2 == 0) continue; // Skip self interaction
      float u = (K * charge.getCharge() * this.charge * r.len()) / (r2 + HEIGHT2);
      electricPotential += u;
    }
    return electricPotential;
  }

  public void draw(ShapeDrawer drawer) {
    if (!fixed) {
      if (drawPath) {
        path.drawPath(drawer);
      }
      if (drawArrow) {
        Vector2 end = screenPosition.cpy().mulAdd(velocity, 1e-6f / Main.SCALE);
        Draw.drawArrow(drawer, screenPosition.cpy(), end.cpy(), Color.BLUE);

        end = screenPosition.cpy().mulAdd(force, 1e15f / Main.SCALE);
        Draw.drawArrow(drawer, screenPosition.cpy(), end.cpy(), Color.RED);
      }
    }

    Color tempColor = color.cpy();
    if (selected) {
      tempColor.mul(SELECTED_LIGHT_FACTOR);
    }
    drawer.filledCircle(screenPosition, RADIUS, tempColor);
  }

  public float getDistanceTraveled() {
    return distanceTraveled;
  }
}
